src/crypto/tls/ech.go

```diff
@@ -882,6 +885,85 @@ func echIsValidVersion(ext uint16) bool {
 	return ext == extensionECH
 }
 
+// encodeClientHelloOuterAAD interprets data as ClientHelloOuter and maps it to
+// ClientHelloOuterAAD. Returns a bit indicated whether parsing ClientHelloOuter
+// succeeded.
+func encodeClientHelloOuterAAD(data []byte, ext uint16) ([]byte, bool) {
+	var (
+		errReadFailure           = errors.New("read failure")
+		msgType                  uint8
+		legacyVersion            uint16
+		random                   []byte
+		legacySessionId          cryptobyte.String
+		cipherSuites             cryptobyte.String
+		legacyCompressionMethods cryptobyte.String
+		extensions               cryptobyte.String
+		s                        cryptobyte.String
+		b                        cryptobyte.Builder
+	)
+
+	u := cryptobyte.String(data)
+	if !u.ReadUint8(&msgType) ||
+		!u.ReadUint24LengthPrefixed(&s) || !u.Empty() {
+		return nil, false
+	}
+
+	if !s.ReadUint16(&legacyVersion) ||
+		!s.ReadBytes(&random, 32) ||
+		!s.ReadUint8LengthPrefixed(&legacySessionId) ||
+		!s.ReadUint16LengthPrefixed(&cipherSuites) ||
+		!s.ReadUint8LengthPrefixed(&legacyCompressionMethods) {
+		return nil, false
+	}
+
+	if !s.Empty() &&
```

Comment:
Shouldn't extensions be non-empty? (Meaning, should fail if `s.Empty() == true` here?)